File: codes/supporting_codes/continuous_stim_waveform.m
function phi_x = continuous_stim_waveform(step, params)
% continuous_stim_waveform
%   Sinusoidal carrier that is ON only during a "pulse window" repeated at PRF.
%
% Time conventions:
%   step is 0-based (same as TMS(step-1,...))
%   t = step * params.deltat
%
% Required fields (recommended under params.stim.cont):
%   params.stim.cont.prf_hz                 % pulse repetition frequency
%   params.stim.cont.pulse_duration_s       % pulse ON duration each PRF period
%   params.stim.cont.sin_freq_hz            % sinusoidal frequency inside pulse
%   params.stim.cont.amp                    % sinusoidal amplitude
%   params.stim.cont.mean                   % DC mean added during pulse
%   params.stim.cont.phase_rad              % optional, default 0
%
% Returns:
%   scalar phi_x

    dt = params.deltat;
    t  = step * dt;

    c = params.stim.cont;

    % Off before onset or after total duration
    if t < params.onset || t >= (params.onset + params.Duration)
        phi_x = 0;
        return;
    end

    % PRF period and time within current PRF cycle
    Tprf = 1 / c.prf_hz;
    tau  = t - params.onset;
    tau_in_period = tau - floor(tau / Tprf) * Tprf;  % modulo without mod() fp quirks

    % Gate: ON only within pulse_duration
    if tau_in_period >= c.pulse_duration_s
        phi_x = 0;
        return;
    end

    % Sinusoid during ON window
    if ~isfield(c,'phase_rad'), c.phase_rad = 0; end
    phi_x = c.mean + c.amp * sin(2*pi*c.sin_freq_hz * tau + c.phase_rad);
end

File: codes/py_codes/continuous_stim_waveform.py
import numpy as np


def continuous_stim_waveform(step, params):
    """
    Python translation of continuous_stim_waveform.m.

    Sinusoidal carrier that is ON only during a pulse window repeated at PRF.

    Important MATLAB convention from continuous_stim_waveform.m:
        step is treated as 0-based, and t = step * params.deltat.

    Returns a scalar phi_x.
    """
    dt = params.deltat
    t = step * dt
    c = params.stim.cont

    # Off before onset or after total duration: [onset, onset + Duration)
    if t < params.onset or t >= (params.onset + params.Duration):
        return 0.0

    Tprf = 1.0 / c.prf_hz
    tau = t - params.onset

    # Same formula as MATLAB: tau - floor(tau/Tprf)*Tprf.
    # This avoids depending directly on np.mod near boundaries.
    tau_in_period = tau - np.floor(tau / Tprf) * Tprf

    if tau_in_period >= c.pulse_duration_s:
        return 0.0

    phase_rad = getattr(c, 'phase_rad', 0.0)
    phi_x = c.mean + c.amp * np.sin(2.0 * np.pi * c.sin_freq_hz * tau + phase_rad)
    return float(phi_x)
